broker-core/src/main/java/io/zeebe/broker/logstreams/processor/TypedStreamProcessor.java

```diff
@@ -159,16 +169,61 @@ public void wrap(
 
     @Override
     public void processEvent() {
-      writer.reset();
-      responseWriter.reset();
-
-      this.writer.configureSourceContext(streamProcessorId, position);
+      resetOutput();
 
       // default side effect is responses; can be changed by processor
       sideEffectProducer = responseWriter;
 
-      eventProcessor.processRecord(
-          position, event, responseWriter, writer, this::setSideEffectProducer);
+      final boolean isNotOnBlacklist = !zeebeState.isOnBlacklist(event);
+      if (isNotOnBlacklist) {
+        eventProcessor.processRecord(
+            position, event, responseWriter, writer, this::setSideEffectProducer);
+      }
+    }
+
+    @Override
+    public void processingFailed(Exception exception) {
+      resetOutput();
+
+      final String errorMessage =
+          String.format(PROCESSING_ERROR_MESSAGE, event, exception.getMessage());
+      LOG.error(errorMessage, exception);
+
+      if (event.metadata.getRecordType() == RecordType.COMMAND) {
+        sendCommandRejectionOnException(errorMessage);
+        writeCommandRejectionOnException(errorMessage);
+      }
+
+      final Intent intent = event.getMetadata().getIntent();
+      if (isWorkflowInstanceRelated(intent)) {
+        if (shouldBeBlacklisted(intent)) {
+          zeebeState.blacklist(event);
+        }
+      }
+    }
+
+    private boolean shouldBeBlacklisted(Intent intent) {
```

Comment:
Can we enforce here that the argument is `WorkflowInstanceRelatedIntent`? In case someone ever calls this and expects us to do the check here. So the caller has to check and cast. 

Or merge both `isWorkflowInstanceRelated` and `shouldBeBlacklisted`, whichever you prefer.